flax/optim.py

```diff
@@ -488,25 +488,21 @@ def _check_inputs(inputs):
           'ModelParamTraversal can only traverse a flax Model instance.')
 
   def _iterate(self, x, path=''):
-    if not isinstance(x, dict):
-      # x is a leaf
-      if self._filter_fn(path, x):
-        yield x
-    else:
-      for key, value in _sorted_items(x):
-        yield from self._iterate(value, '{}/{}'.format(path, key))
+    flat_dict = traverse_util.flatten_dict(x)
+    for key, value in _sorted_items(flat_dict):
+      path = '/' + '/'.join(key)
```

Comment:
Probably not worth changing here but wouldn't it be nicer to have the flattened keys be actual arrays of keys, e.g. `['foo', 'bar']` instead of `"/foo/bar"`?

Reply:
`flatten_dict` uses a tuple of strings `('foo', 'bar')` because arrays are not hashable but yes we could consider using the same syntax for the traversals. The downside is that pattern matching on a tuple of strings is a little harder. I will not do it in this PR anyway because it would be a breaking change

Reply:
Ah yes, tuple is better than array.